src/IO/H5/CombineH5.cpp

```diff
@@ -149,4 +157,186 @@ void combine_h5_vol(const std::vector<std::string>& file_names,
+      std::optional<size_t> row = times.rows() - 1;
+      while (times(row.value(), 0) >= earliest_time.value()) {
+        // Once we get to the first time in this file, if it's still after the
+        // earliest time in the previous file, then we don't include any times
+        // from this file.
+        if (row.value() == 0) {
+          row.reset();
+          break;
+        }
+        row.value()--;
+      }
+
+      // So long as this file contains some times that need to be combined,
+      // store the number of times and the first time in this file.
+      if (row.has_value()) {
+        num_time_map.at(dat_filename)[index] = row.value() + 1;
+        earliest_time = times(0, 0);
+      }
+
+      file_to_combine.close_current_object();
+    }
+  }
+
+  if (verbosity >= Verbosity::Verbose) {
+    std::stringstream ss{};
+    ss << "Number of times selected to combine in each H5 file for each dat "
+          "subfile:\n";
+    for (const auto& [subfile_name, h5_files_num_times] : num_time_map) {
+      ss << " Dat Subfile " << subfile_name << ":\n";
+      for (size_t i = 0; i < h5_files_num_times.size(); i++) {
+        ss << "  H5 File " << h5_files_to_combine[i] << ": "
+           << h5_files_num_times[i].value_or(0) << "\n";
+      }
+    }
+
+    Parallel::printf("%s", ss.str());
+  }
+
+  // Now that we know the time indices for each dat file for each H5 file to
+  // combine, we open the output file and combine everything
+  h5::H5File<h5::AccessType::ReadWrite> output_h5_file{output_h5_filename};
+
+  const auto make_these_columns = [](const size_t size) -> std::vector<size_t> {
+    std::vector<size_t> these_columns{};
+    these_columns.resize(size);
+    alg::iota(these_columns, 0_st);
+    return these_columns;
+  };
+
+  // Now we loop over H5 files first to avoid unnecessary filesystem access
+  for (size_t i = 0; i < h5_files_to_combine.size(); i++) {
+    const std::string& filename = h5_files_to_combine[i];
+    const h5::H5File<h5::AccessType::ReadOnly> file_to_combine{filename};
+    for (const std::string& dat_filename : subfile_dat_names) {
+      const auto& input_dat_file = file_to_combine.get<h5::Dat>(dat_filename);
+      const std::vector<std::string>& legend = input_dat_file.get_legend();
+
+      // Avoid copying the legend around if we don't have to
+      h5::Dat* output_dat_file = nullptr;
+      if (output_h5_file.exists<h5::Dat>(dat_filename)) {
+        output_dat_file = &output_h5_file.get<h5::Dat>(dat_filename);
+      } else {
+        const uint32_t version = input_dat_file.get_version();
+        output_dat_file =
+            &output_h5_file.insert<h5::Dat>(dat_filename, legend, version);
+      }
+
+      const std::optional<size_t>& num_times = num_time_map.at(dat_filename)[i];
+
+      // Only append data if we include data from this file
+      if (num_times.has_value()) {
+        // Always start with row 0
+        const Matrix data_to_append = input_dat_file.get_data_subset(
```

Comment:
This will need to load the entire matrix, sort it, and then trim it before write.